From f34e9d8e9a34c6b843e4be6ecf641a3b8aece2a8 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Thu, 13 Apr 2017 14:01:53 +0200
Subject: [PATCH] OS-6031 Need NVMe dump support Reviewed by: Robert Mustacchi
 <rm@joyent.com> Reviewed by: Jerry Jelinek <jerry.jelinek@joyent.com>
 Approved by: Robert Mustacchi <rm@joyent.com>

---
 usr/src/uts/common/io/nvme/nvme.c | 22 +++++++++++++++-------
 1 file changed, 15 insertions(+), 7 deletions(-)

diff --git a/usr/src/uts/common/io/nvme/nvme.c b/usr/src/uts/common/io/nvme/nvme.c
index cb2e9bdd22..c7bf900fe0 100644
--- a/usr/src/uts/common/io/nvme/nvme.c
+++ b/usr/src/uts/common/io/nvme/nvme.c
@@ -2964,26 +2964,34 @@ static int
 nvme_bd_cmd(nvme_namespace_t *ns, bd_xfer_t *xfer, uint8_t opc)
 {
 	nvme_t *nvme = ns->ns_nvme;
-	nvme_cmd_t *cmd;
+	nvme_cmd_t *cmd, *ret;
+	nvme_qpair_t *ioq;
 
 	if (nvme->n_dead)
 		return (EIO);
 
-	/* No polling for now */
-	if (xfer->x_flags & BD_XFER_POLL)
-		return (EIO);
-
 	cmd = nvme_create_nvm_cmd(ns, opc, xfer);
 	if (cmd == NULL)
 		return (ENOMEM);
 
 	cmd->nc_sqid = (CPU->cpu_id % nvme->n_ioq_count) + 1;
 	ASSERT(cmd->nc_sqid <= nvme->n_ioq_count);
+	ioq = nvme->n_ioq[cmd->nc_sqid];
 
-	if (nvme_submit_cmd(nvme->n_ioq[cmd->nc_sqid], cmd)
-	    != DDI_SUCCESS)
+	if (nvme_submit_cmd(ioq, cmd) != DDI_SUCCESS)
 		return (EAGAIN);
 
+	if ((xfer->x_flags & BD_XFER_POLL) == 0)
+		return (0);
+
+	do {
+		ret = nvme_retrieve_cmd(nvme, ioq);
+		if (ret != NULL)
+			nvme_bd_xfer_done(ret);
+		else
+			drv_usecwait(10);
+	} while (ioq->nq_active_cmds != 0);
+
 	return (0);
 }
 
-- 
2.21.0

